File: generated-code/python/codecraft/TcpReadWrite/model/build_action.py
from model.entity_type import EntityType
from stream_wrapper import StreamWrapper
from vec2_int import Vec2Int

class BuildAction:
    """Build action"""

    __slots__ = ("entity_type","position",)

    entity_type: EntityType
    position: Vec2Int

    def __init__(self, entity_type: EntityType, position: Vec2Int):
        self.entity_type = entity_type
        """Type of an entity to build"""
        self.position = position
        """Desired position of new entity"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "BuildAction":
        """Read BuildAction from input stream
        """
        entity_type = EntityType(stream.read_int())
        position = Vec2Int.read_from(stream)
        return BuildAction(entity_type, position)
    
    def write_to(self, stream: StreamWrapper):
        """Write BuildAction to output stream
        """
        stream.write_int(self.entity_type)
        self.position.write_to(stream)
    
    def __repr__(self):
        return "BuildAction(" + \
            repr(self.entity_type) + \
            ", " + \
            repr(self.position) + \
            ")"

File: generated-code/haskell/codecraft/TcpReadWrite/src/Model/BuildAction.hs
module Model.BuildAction where

import Prelude hiding (id)
import qualified Trans
import Trans (Trans)
import Model.EntityType (EntityType)
import Vec2Int32 (Vec2Int32)

-- | Build action
data BuildAction = BuildAction {
    -- | Type of an entity to build
    entityType :: EntityType,
    -- | Desired position of new entity
    position :: Vec2Int32 }
    deriving Show

instance Trans BuildAction where
    read = do
        entityType <- Trans.read
        position <- Trans.read
        return BuildAction {
            entityType,
            position }

    write BuildAction {
        entityType,
        position } = do
            Trans.write entityType
            Trans.write position